[Docs][ES|QL] Trim remote cluster setup from CCS page (#149314)
elastic/elasticsearch: docs/reference/query-languages/esql/esql-cross-clusters.md

Reviewer: I think the entire `## prerequisites` section needs to be recreated, as it's a copy & pase from [Cross Cluster Search](https://www.elastic.co/docs/explore-analyze/cross-cluster-search) (we will have to improve that other document in a different PR too)

If this doc is for `ES|QL across clusters`,  we need to state what ES|QL across clusters requires specifically, not what {{ccs-cap}} requires, and if it requires the same (which I doubt as standard CCS supports both API key or TLS certs authentication while ES|QL apparently only supports API key based auth) we should just mention that we require the same as CCS.

Based on what I've seen ES|QL requires Remote Clusters to be setup with API authentication as the security model.

IMO we shouldn't mention anything about sniff mode or proxy mode, as those are already requirements of remote clusters configuration, and feels completely out of the scope in this document.

What we have to include is what exact requirements (and for example the ~API key~ **role example** with permissions gives value here) ES|QL has besides having 2 clusters properly connected through remote clusters.

## Prerequisites [esql-ccs-prerequisites]

> * {{esql}} {{ccs}} is not supported in {{serverless-short}}.

Reviewer: The serverless statement doesn't feel like a prerequisite. You have a serverless note in the introduction of the document, just make sure that it covers this sentence about ES|QL CCS not supported in serverless.

And I'd say that more or less it already covers it so we could just remove this sentence.

> * For full {{ccs}} capabilities, the local and remote clusters must be on the same [subscription level](https://www.elastic.co/subscriptions).

Reviewer: Doesn't this feature require Enterprise?
<img width="1512" height="173" alt="image" src="https://github.com/user-attachments/assets/40b79ffb-26e8-456a-8a9e-7213c6449230" />

Also API Key based auth for remote clusters require enterprise subscription, so saying "the same" feels odd. I think we should say that `ES|QL through CCS requires an Enterprise subscription level in both local and remote clusters`.

Maybe we have to confirm that with PMs but I think it's clear.

* The local coordinating node must have the [`remote_cluster_client`](docs-content://deploy-manage/distributed-architecture/clusters-nodes-shards/node-roles.md#remote-node) node role.
* The remote clusters must be connected using **API key authentication**. For setup instructions across all deployment types, refer to [Add remote clusters using API key authentication](docs-content://deploy-manage/remote-clusters/remote-clusters-api-key.md).
  

>   To verify which security model is in use, run `GET _remote/info`. With API key authentication, the response includes a `"cluster_credentials"` key.

Reviewer: ```suggestion
- [Remote clusters](docs-content://deploy-manage/remote-clusters.md) must be configured before running {{esql}} across clusters.

  The destination cluster must be configured as a remote cluster on the local cluster, and the remote cluster connection must be correctly established. For setup instructions, refer to [Set up remote clusters](docs-content://deploy-manage/remote-clusters.md#setup).

- {{esql}} across clusters requires the remote cluster connection to use the [API key–based security model](docs-content://deploy-manage/remote-clusters/security-models.md#api-key).

  To verify which security model is active, run `GET _remote/info`. When API key authentication is in use, the response includes `"cluster_credentials"`.
```

Consider this way to present the remote clusters requirement, in two different bullets.

In the first one we explicitly mention and link to the remote clusters landing page and "setup" for all deployment types. We clearly state the goal (connection correctly established before moving forward).

In the second bullet we explicitly mention the API key based auth with a link to the description of the security model. I don't include a setup link here because the setup link on the previous bullet should include all deployment types and cases. This second bullet also states the goal (verify / ensure you are using API key based auth with `GET _remote/info`).

## Prerequisites [esql-ccs-prerequisites]

> * {{esql}} {{ccs}} requires an [Enterprise subscription](https://www.elastic.co/subscriptions) on both the local (querying) cluster and every remote cluster. The legacy TLS certificate-based CCS model available at lower tiers is deprecated in 9.0 and does not support {{esql}}.

Reviewer: Talking about "the legacy TLS certificate-based model" on the first pre-requisite feels unexpected and difficult to understand. The requisite talks about a license level and the second sentence talks about a low-level implementation method of remote clusters, but not mentioned at all.

I'd probably remove that second sentence and leave just the enterprise subscription requirement. You can consider that sentence later on where you add the API key based auth model, but it would feel repetitive as saying it "requires" api key based auth is the same as saying TLS cert-based auth is not supported :)

> * The local coordinating node must have the [`remote_cluster_client`](docs-content://deploy-manage/distributed-architecture/clusters-nodes-shards/node-roles.md#remote-node) node role.

Reviewer: I'd remove this sentence about `remote_cluster_client` role, or include it as sub-item of the remote clusters stuff. But IMO this is already required by remote cluster connections and how remote cluster connections work.

Reviewer: If we want to keep that sentence I'd refine it a bit to be more friendly, like:

* The local node receiving the query must have the [`remote_cluster_client`](docs-content://deploy-manage/distributed-architecture/clusters-nodes-shards/node-roles.md#remote-node) node role to be able to connect to the remote clusters.

And add it **after** the remote clusters requirement, as it relates with remote clusters.

    To ensure your remote cluster configuration supports {{ccs}}, see [Supported {{ccs}} configurations](docs-content://explore-analyze/cross-cluster-search.md#ccs-supported-configurations).

* For full {{ccs}} capabilities, the local and remote clusters must be on the same [subscription level](https://www.elastic.co/subscriptions).
* The local coordinating node must have the [`remote_cluster_client`](docs-content://deploy-manage/distributed-architecture/clusters-nodes-shards/node-roles.md#remote-node) node role.

> * If you use [sniff mode](docs-content://deploy-manage/remote-clusters/remote-clusters-self-managed.md#sniff-mode), the local coordinating node must be able to connect to seed and gateway nodes on the remote cluster.

Reviewer: sniff mode requirement should be removed from this doc, that's internal to remote clusters, not specific to ES|QL, neither CCS. Feels too low level and out of place here. If needed we could add it to the remote clusters docs.

    We recommend using gateway nodes capable of serving as coordinating nodes. The seed nodes can be a subset of these gateway nodes.

> * If you use [proxy mode](docs-content://deploy-manage/remote-clusters/remote-clusters-self-managed.md#proxy-mode), the local coordinating node must be able to connect to the configured `proxy_address`. The proxy at this address must be able to route connections to gateway and coordinating nodes on the remote cluster.

Reviewer: Same, proxy mode requirement should be removed from here.

Of course we could say in this doc that both proxy and sniff modes are supported, with links to the remote cluster docs that describe the modes, so the user gets more context.

But IMO we should just point to remote clusters and ensure the reader understands that before being able to play with ES|QL across clusters then need to have 2 clusters correctly interconnected through "remote clusters", that's all.

> * {{ccs-cap}} requires different security privileges on the local cluster and remote cluster. See [Configure roles and users](docs-content://deploy-manage/remote-clusters/remote-clusters-api-key.md#remote-clusters-privileges-api-key) and [*Remote clusters*](docs-content://deploy-manage/remote-clusters.md).

Reviewer: I don't see the value of that paragraph either, considering we are offering a role and user in the next section (API key authentication).

I'd redo the entire prerequisites saying something like:

> ES|QL requires remote clusters to be configured, with API key authentication as the security model. Refer to xxxx and xxxx for details on how to configure remote clusters for all deployment types.

Then we can proceed giving an example setup of users and roles for the examples to work, and we can offer a link to "Configure users and roles" (`/deploy-manage/remote-clusters/remote-clusters-api-key.md#_configure_privileges_for_ccs`) there.

> ## API key authentication [esql-ccs-security-model-api-key]

Reviewer: ```suggestion
## Configure user and role
```

The section is about configuring a role with CCS privileges and a user.

API key authentication is the pre-requisite, this section does not talk about API key authentication, it assumes remote cluster is setup with API key authentication and proceed to suggest the creation of a role and a user for the next examples to work.

I'd change the title only, the content is accurate and needed for this doc.

```

1. The `cross_cluster_search` cluster privilege is required for the *local* cluster.

> 2. Typically, users have permissions to read both local and remote indices. However, for cases where the role is intended to *only* search the remote cluster, the `read` permission is still required for the local cluster. To provide read access to the local cluster, but disallow reading any indices in the local cluster, the `names` field may be an empty string.

Reviewer: That footnote `2.` feels a bit unexpected. Is this an ES|QL need or something global for CCS?

Is there any better explanation that we could provide so a reader doesn't feel this like a `hack`? It would be great to explain why we require a minimum local privilege of `read` towards `""`. 🤷 

Anyway the text is super clear, so that's ok and we can ignore my comment.

Maybe @quux00 can give us some extra info about that requirement.

Author: You said we could ignore this one, but I couldn't resist looking. Turns out it's an ES|QL-specific thing, not a global CCS one: ES|QL authorizes the query at the local coordinating node by action name before resolving which clusters it will target, so the role still needs a local `read` grant even for purely remote queries. Classic CCS goes through per-resolved-index authorization, so it doesn't hit this branch. Updated the footnote to explain the why. @quux00 might have more context if I got anything wrong.

Reviewer: thanks a lot for such a nice explanation! I'd then add ES|QL somewhere in the sentence then, let me try a small suggestion :)

3. The indices allowed read access to the remote cluster. The configured [cross-cluster API key](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-security-create-cross-cluster-api-key) must also allow this index to be read.
4. The `read_cross_cluster` privilege is always required when using {{esql}} across clusters with the API key based security model.

> 5. The remote clusters to which these privileges apply. This remote cluster must be configured with a [cross-cluster API key](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-security-create-cross-cluster-api-key) and connected to the local cluster before the remote index can be queried. Verify connection using the [Remote cluster info](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-cluster-remote-info) API.

Reviewer: ```suggestion
5. The remote clusters to which these privileges apply. This remote cluster must be configured with a [cross-cluster API key](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-security-create-cross-cluster-api-key) and connected to the local cluster before the remote index can be queried. Verify connection using the [Remote cluster info](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-cluster-remote-info) API. For setup instructions, refer to [Set up remote clusters](docs-content://deploy-manage/remote-clusters.md#setup).
```

Not sure if the footnote ends up being too long. Feel free to ignore.

Author: Thanks for the thought here! I'm going to skip this one : the new prerequisites bullet up top now links to Set up remote clusters, so adding it again in the footnote would point readers to the same page twice. And as you noted, that keeps the footnote a bit lighter too. Happy to revisit if you'd rather have it in both spots! :)

```

1. The `cross_cluster_search` cluster privilege is required for the *local* cluster.

> 2. {{esql}} authorizes the query at the local coordinating node before resolving which clusters it will target. A role used only for remote queries must therefore still grant `read` in the local `indices` block. Setting `names` to an empty string satisfies that check while matching no real index, so the user cannot read any local data. This requirement is specific to {{esql}}; classic {{ccs}} authorizes per resolved index and does not need a local grant.

Reviewer: ```suggestion
2. {{esql}} authorizes the query at the local node before resolving which clusters it will target. A role used only for remote queries must therefore still grant `read` in the local `indices` block. Setting `names` to an empty string satisfies that check while matching no real index, so the user cannot read any local data. This requirement is specific to {{esql}}; classic {{ccs}} authorizes per resolved index and does not need a local grant.
```

Super nice explanation!

I prefer to remove `coordinating` from this sentence as the majority of users don't really know what we mean with that.

What we mean in this exact context is the node receiving the query, which acts as coordinating-node for the execution of the query.

In Elasticsearch all nodes can act as coordinating-nodes, no matter the roles they have. But mentioning it too much is not going to help readers. Also coordinating-only nodes are nodes with no roles associated (they can just coordinate indexing & search requests, they don't store any data and they don't act as masters). That's a more known concept and could potentially cause confusion here if we mention `coordinating nodes` too much without scoping it better.

I'd leave `local node` or `local node receiving the query`.

Reviewer: Or `the node coordinating the query execution` :)

}
```

> All cross-cluster requests from the local cluster are bound by the cross-cluster API key’s privileges, which are controlled by the remote cluster’s administrator.

Reviewer: ```suggestion
::::{note}
All cross-cluster requests from the local cluster are bound by the cross-cluster API key’s privileges, which are controlled by the remote cluster’s administrator. Local roles can only further reduce these permissions, they cannot increase access beyond what the API key allows.
::::{note}
```

Not sure if in a note or in the way it is, but we should enrich that sentence, which is very important.

The key is for the user to understand that both the created role and the cross-cluster API key (used during remote clusters setup) are important for the final effective permissions of the user.

If for example the admin of the remote cluster created an API key without access to the needed indices, then the role we create in this section is useless. That's why maybe a note is interesting here.

For your consideration.

> ## Query across multiple clusters [ccq-from]

Reviewer: I'd make an intro for this section and set expectations.

For example we should say that in the following examples `cluster_one` and `cluster_two` represent remote clusters configured in the local cluster where the queries are being executed.

Probably there's no need to link to remote clusters docs anymore as we have done at the beginning of the doc.

But we need to be sure that a reader understands `cluster_one` and maps it mentally as a remote cluster registered in the local cluster (as we said before).
